### PR TITLE
AC: Treat Directories as blobs

### LINKED ISSUE #33: PushBlob broken on blobs that unmarshal as `Directory`s using the Action Cache asset store
Due to the API design we have, we don't pass through whether the object is supposed to be a directory, so in order to construct an `ActionResult` with the correct semantics we must download the referenced asset from CAS, and see if it unmarshals as a `Directory`.  If the thing pushed is a blob that parses as a `Directory`, we treat it as if it were one, and later probably fail due to the `Directory` we have being corrupted.  BuildStream's [`Source` proto](https://github.com/apache/buildstream/blob/master/src/buildstream/_protos/buildstream/v2/source.proto) is such a problematic message.

Note that what we're currently doing isn't quite in line with the spec anyway, as there's not actually a requirement that `PushDirectory` references a `Directory` proto.

### PR DESCRIPTION
**Pull request by tomcoldrick-ct**
We currently do some cleverness to pop `Directory` objects into the `OutputDirectories` field of the `ActionResult` for semantic purposes. Unfortunately doing this requires jumping over many hurdles, partially of our own making and partially from REAPI.

To detect whether the object is a `Directory`, we fetch it from CAS and attempt to unmarshal it into a `Directory` message.  We then must convert it to a `Tree` to store in the `ActionResult`, which also requires the raw proto.

The catch is that things that are not `Directory` messages may successfully unmarshal into one -- as a motivating example, BuildStream's [`Source` proto](https://github.com/apache/buildstream/blob/master/src/buildstream/_protos/buildstream/v2/source.proto) is such a message.  When this happens, we're very likely to fail the request as we attempt to use the corrupted `Directory` we've created.  In the BuildStream case, we hit an error with "Unknown Digest Function", as a `Digest` proto is loaded as `nil`.

To solve this, we could pass through data on whether the asset is supposed to be a directory or not, however doing so is inelegant and has another subtle problem -- if the client decides to use `PushDirectory` to push something that isn't a `Directory` (which isn't explicitly banned by the spec), then we will error.

As the `Action` and `ActionResult`s we generate are intended only to be used by us, we can instead break the semantics somewhat and treat the `Directory` as an opaque file, as we do for blobs.

Note: this intentionally breaks the sharing of `Action`s should the `RemoteExecutionFetcher` be used.  This feels more correct to me, as we were overwriting the one we actually ran anyway.  Under this usage we instead get multiple `Action`s pointing to the same `ActionResult` -- one for the actually executed `Action`, and another for the asset reference.

Fixes #33

### CODE REVIEW

**Comment by EdSchouten** [comment]
Hey! Before merging any of this, take a look at this PR:

https://github.com/bazelbuild/remote-apis/pull/258

bb-storage already implements this:

https://github.com/buildbarn/bb-storage/commit/068d214ed6a3c10ff405e3d72d9a27710b0256cc

What you need to do is write Action Cache entries with OutputDirectories that have both `tree_digest` and `root_directory_digest` set. So upload directories both as a Tree and separate Directory messages. Once you do that, CompletenessCheckingBlobAccess is capable of using the Tree message to ensure the separate Directory messages also remain present.

In other words, you won't need to abuse using OutputFiles for this. Hope that helps!

**Comment by harrysarson** [comment]
I am taking over this work. What I would like to do is

1. Add a field to the remote-asset internal `Asset` proto to indicate whether this is a "blob" or a "directory". This field would control whether the `ActionResult` uses `output_files` or `output_directories` to store the asset's digest (instead of our current heuristic of loading the blob from the CAS and seeing what it looks like).
   - This field could also live on the `AssetReference` proto, would need to play about a bit with this.
2. In the action-cache asset store `Put`
   - Check to `Asset` to determine if this is "blob" or "directory"
   - Set `output_directory_format = DIRECTORY_ONLY` in the `Command` we create (for directory assets).
   - Set `output_directories.root_directory_digest` (instead of setting `output_directories.tree_digest`) in the `ActionResult` we create (for directory assets).
3. In the action-cache asset store `Get`
   - Use the `output_directories.root_directory_digest` from the `ActionResult` as the digest to return to the client.
   - If the output_directories digest is empty then this is a "blob" asset.
  
This uses the `output_directory_format` added in https://github.com/bazelbuild/remote-apis/pull/258.

File: pkg/storage/action_cache_asset_store.go [hunk 1]
```diff
@@ -204,22 +156,10 @@ func (rs *actionCacheAssetStore) Put(ctx context.Context, ref *asset.AssetRefere
...
-			return err
-		}
-		action = &remoteexecution.Action{
-			CommandDigest:   commandDigest,
-			InputRootDigest: EmptyDigest,
-		}
+	// Create the action with the qualifier directory as the input root
+	action, command, err = assetReferenceToAction(ref, directoryDigest)
+	if err != nil {
+		return err
```
Comment (harrysarson) [inline pkg/storage/action_cache_asset_store.go:162]: This change makes assets with the special qualifiers we check for in `QualifiersToCommand` to unfetchable from the asset server.

Either we need to revert this bit or apply the corresponding change (i.e dropping `QualifiersToCommand`) from `Get`.

**Comment by harrysarson** [comment]
> What you need to do is write Action Cache entries with OutputDirectories that have both tree_digest and root_directory_digest set.

I do not believe its feasible to set the `tree_digest` when writing the Action Cache entry. When creating the entry all we know about the asset is its digest (we also know the uris and qualifiers but they don't help us). Getting a `tree_digest` requires doing "assume digest refers to a directory and look it up in the CAS" which is what we are trying to avoid doing.

So I think we need to write Action Cache entries with OutputDirectories that **only** have root_directory_digest set.

**Comment by tomcoldrick-ct** [comment]
> > What you need to do is write Action Cache entries with OutputDirectories that have both tree_digest and root_directory_digest set.
> 
> I do not believe its feasible to set the `tree_digest` when writing the Action Cache entry. When creating the entry all we know about the asset is its digest (we also know the uris and qualifiers but they don't help us). Getting a `tree_digest` requires doing "assume digest refers to a directory and look it up in the CAS" which is what we are trying to avoid doing.
> 
> So I think we need to write Action Cache entries with OutputDirectories that **only** have root_directory_digest set.

If we're told that the asset is _supposed_ to be a `Directory`, by the endpoint being called being `{Push,Fetch}Directory`, then it's perfectly valid to attempt to pull and form a `Tree` from it.  If the thing pointed at isn't a valid directory, then that is an error for the client.  If we store whether the pointed-to-thing is a `Directory` or not on the `Asset`, then I think it's reasonable to return an error if we try to fetch something that's not a `Directory` with `FetchDirectory`.

**Comment by harrysarson** [comment]
> I think it's reasonable to return an error if we try to fetch something that's not a `Directory` with `FetchDirectory`.

In #33 you wrote:

> Note that what we're currently doing isn't quite in line with the spec anyway, as there's not actually a requirement that PushDirectory references a Directory proto.

Have we changed our mind on this?

---

Either way, calculating the `tree_digest` requires recursively reading from the CAS which feels like something that `assetStore.Put` should avoid doing if it can.

**Comment by EdSchouten** [comment]
>    * Set `output_directory_format = DIRECTORY_ONLY` in the `Command` we create (for directory assets).
>    * Set `output_directories.root_directory_digest` (instead of setting `output_directories.tree_digest`) in the `ActionResult` we create (for directory assets).

Note that this is something that bb-storage's CompletenessCheckingBlobAccess does not support. You must set `tree_digest`. `root_directory_digest` may be set in addition to that. The reason being that bb-storage uses Tree objects to efficiently obtain a list of all recursively referenced Directory objects, so that their presence can be checked.
